File: bindings/ruby/test/test_leak.rb
# C9 leak-discipline suite for the Ruby facade, measured with GC + ObjectSpace.
# A >=10,000-cycle create/use/free loop must reach a steady state: after a full
# GC no facade wrapper (Transducer, Query) may remain live beyond the warmup
# baseline, and the total live-object count must not drift upward. Native
# handles are freed at close() (and by finalizers); ObjectSpace confirms the
# managed wrappers are reclaimed rather than accumulating.

require "minitest/autorun"
require "vinary_tree/liblevenshtein"
require "vinary_tree/libdictenstein"

class LeakTest < Minitest::Test
  M = VinaryTree::Liblevenshtein
  D = VinaryTree::Libdictenstein
  CYCLES = 10_000
  WARMUP = 2_000
  # Generous slot-growth ceiling; a per-cycle leak would accrue far more.
  MAX_SLOT_GROWTH = 200_000

  def full_gc
    GC.start(full_mark: true, immediate_sweep: true)
  end

  def live_count(klass)
    full_gc
    ObjectSpace.each_object(klass).count
  end

  def transducer_cycle
    dictionary = D::DynamicDawg.new
    %w[cat cot cut scat].each_with_index { |term, index| dictionary.put(term, index) }
    transducer = M::Transducer.new(dictionary)
    transducer.query("cat", 2).each { |_match| }
    transducer.close
    dictionary.close
  end

  def test_transducer_cycle_reclaims_wrappers_and_holds_steady
    WARMUP.times { transducer_cycle }
    baseline_transducers = live_count(M::Transducer)
    baseline_queries = live_count(M::Query)
    full_gc
    baseline_slots = GC.stat(:heap_live_slots)

    CYCLES.times { transducer_cycle }

    assert_operator live_count(M::Transducer), :<=, baseline_transducers,
                    "Transducer wrappers accumulated across #{CYCLES} cycles"
    assert_operator live_count(M::Query), :<=, baseline_queries,
                    "Query wrappers accumulated across #{CYCLES} cycles"
    full_gc
    growth = GC.stat(:heap_live_slots) - baseline_slots
    assert_operator growth, :<, MAX_SLOT_GROWTH,
                    "live slots grew #{growth} across #{CYCLES} cycles"
  end

  def test_phonetic_cycle_holds_steady
    cycle = lambda do
      pattern = M::PhoneticPattern.compile_regex("c[ao]t")
      pattern.matches?("cat")
      pattern.close
      rules = M::PhoneticRuleSet.builtin(M::PhoneticRuleSet::ENGLISH_ORTHOGRAPHY)
      rules.apply("phone")
      rules.close
    end
    WARMUP.times { cycle.call }
    full_gc
    baseline_slots = GC.stat(:heap_live_slots)
    CYCLES.times { cycle.call }
    full_gc
    growth = GC.stat(:heap_live_slots) - baseline_slots
    assert_operator growth, :<, MAX_SLOT_GROWTH,
                    "live slots grew #{growth} across #{CYCLES} cycles"
  end
end

File: bindings/fortran/test/test_leak.f90
! C9 leak-discipline suite for the Fortran facade. Fortran has no garbage
! collector, so this exercises the create/use/free discipline directly: a
! >=10,000-cycle loop compiles a phonetic pattern, matches against it, and
! closes it each cycle. Resident memory is read from /proc/self/statm before and
! after the measured loop (after a warmup so the allocator has reached a working
! set); a per-cycle native handle leak would push resident pages up without
! bound, whereas correct release keeps them flat within a generous ceiling.
program test_leak
  use, intrinsic :: iso_c_binding, only: c_int32_t, c_long
  use vinary_tree_liblevenshtein
  implicit none

  integer, parameter :: cycles = 10000
  integer, parameter :: warmup = 2000
  ! 4096 pages (~16 MiB) is generous against allocator noise; a real per-cycle
  ! leak would accrue far more over 10k cycles.
  integer(c_long), parameter :: max_growth_pages = 4096_c_long

  integer :: iteration
  integer(c_long) :: baseline, growth

  do iteration = 1, warmup
    call phonetic_cycle()
  end do
  baseline = resident_pages()
  do iteration = 1, cycles
    call phonetic_cycle()
  end do
  growth = resident_pages() - baseline
  if (growth > max_growth_pages) error stop "phonetic cycle leaked resident memory"

  print *, "Fortran leak loop completed cycles"

contains

  subroutine phonetic_cycle()
    type(phonetic_pattern) :: pattern
    integer(c_int32_t) :: status
    logical :: accepted
    call compile_phonetic_regex(pattern, "c[ao]t", status)
    if (status /= llev_ok) error stop "pattern compilation"
    accepted = pattern%matches("cat", status)
    if (status /= llev_ok .or. .not. accepted) error stop "pattern match"
    call pattern%close()
  end subroutine phonetic_cycle

  function resident_pages() result(pages)
    integer(c_long) :: pages
    integer :: unit, ios
    integer(c_long) :: total
    pages = -1_c_long
    open (newunit=unit, file="/proc/self/statm", status="old", action="read", iostat=ios)
    if (ios /= 0) return
    read (unit, *, iostat=ios) total, pages
    close (unit)
  end function resident_pages

end program test_leak
